pkg/controller/elasticsearch/driver/upgrade_deletion_strategy.go

```diff
@@ -0,0 +1,299 @@
+		// This is to prevent a situation where MaxUnavailable is reached and we
+		// can't make some progress even if the user has updated the spec.
+		"do_not_restart_healthy_node_if_MaxUnavailable_reached": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if maxUnavailableReached && k8s.IsPodReady(candidate) {
+				return false, nil
+			}
+			return true, nil
+		},
+		// One master at a time
+		"one_master_at_a_time": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+			// If candidate is not healthy we want give it a chance to restart
+			_, healthy := d.healthyPods[candidate.Name]
+			if !healthy {
+				return true, nil
+			}
+
+			for _, pod := range expectedDeletions {
+				if label.IsMasterNode(pod) {
+					return false, nil
+				}
+			}
+			// Get the expected masters
+			expectedMasters := len(d.masterNodesNames)
+			// Get the healthy masters
+			var healthyMasters []corev1.Pod
+			for _, pod := range d.healthyPods {
+				if label.IsMasterNode(pod) {
+					pod := pod
+					healthyMasters = append(healthyMasters, pod)
+				}
+			}
+			return len(healthyMasters) == expectedMasters, nil
+		},
+		// Force an upgrade of all the data nodes before upgrading the last master
+		"do_not_delete_last_master_if_datanodes_are_not_upgraded": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+			for _, pod := range d.toUpdate {
+				if candidate.Name == pod.Name {
+					continue
+				}
+				if label.IsMasterNode(pod) {
+					// There's some others master alive, allow this one to be deleted
+					return true, nil
+				}
+			}
+			// This is the last master, check if all data nodes are up to date
+			for _, pod := range d.toUpdate {
+				if label.IsDataNode(pod) {
+					// There's still a data node to update
+					return false, nil
+				}
+			}
+			return true, nil
+		},
+		"do_not_delete_last_healthy_master": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+
+			// If only one master node is expected this cluster is not H.A.
+			if len(d.masterNodesNames) < 2 {
+				return true, nil
+			}
+
+			var healthyMasters []corev1.Pod
+			candidateIsHealthy := false
+			for _, expectedMaster := range d.masterNodesNames {
+				for healthyPodName, healthyPod := range d.healthyPods {
+					if !label.IsMasterNode(healthyPod) {
+						continue
+					}
+					master := types.NamespacedName{
```

Comment:
Why do we need a `NamespacedName` here? Looks like you are only interested in the name